src/schnorr.rs

```diff
@@ -0,0 +1,107 @@
+use super::*;
+use ark_ec::{
+    AffineCurve,     // for prime_subgroup_generator()
+    ProjectiveCurve, // for into_affine()
+};
+use ark_ff::{
+    BigInteger, // for is_even()
+    Field,      // for from_random_bytes()
+    PrimeField, // for from_repr()
+    Zero,
+};
+use blake2::{
+    digest::{Update, VariableOutput},
+    VarBlake2b,
+};
+use std::ops::Neg;
+
+pub struct Schnorr<SC: SpongeConstants> {
+    pub sponge: ArithmeticSponge<PallasField, SC>,
+    pub network_id: NetworkId,
+}
+
+impl<SC: SpongeConstants> Signer for Schnorr<SC> {
+    fn sign<I: Input>(&mut self, kp: Keypair, input: I) -> Signature {
+        let k: PallasScalar = self.blinding_hash(&kp, input);
+        let r: PallasPoint = PallasPoint::prime_subgroup_generator().mul(k).into_affine();
+        let k: PallasScalar = if r.y.into_repr().is_even() { k } else { -k };
+
+        let e: PallasScalar = self.message_hash(&kp.pub_key, r.x, input);
+        let s: PallasScalar = k + e * kp.sec_key;
+
+        Signature::new(r.x, s)
+    }
+
+    fn verify<I: Input>(&mut self, sig: Signature, pub_key: PubKey, input: I) -> bool {
+        let ev: PallasScalar = self.message_hash(&pub_key, sig.rx, input);
+
+        let sv: PallasPoint = PallasPoint::prime_subgroup_generator()
+            .mul(sig.s)
+            .into_affine();
+        let rv: PallasPoint = sv + pub_key.mul(ev).neg().into_affine();
+
+        !rv.infinity && rv.y.into_repr().is_even() && rv.x == sig.rx
+    }
+}
+
+impl<SC: SpongeConstants> Schnorr<SC> {
+    pub fn new(sponge: ArithmeticSponge<PallasField, SC>, network_id: NetworkId) -> Self {
+        Schnorr::<SC> { sponge, network_id }
```

Comment:
No, because we need flexibility to specify different sponges for different signing approaches.  For example mainnet uses `ArithmeticSponge::<PallasField, PlonkSpongeConstantsBasic>::new(pasta::fp::params())` but notary signatures will use `ArithmeticSponge::<PallasField, PlonkSpongeConstants5W>::new(pasta::fp5::params())`.